File: src/main/java/fi/quanfoxes/Parser/patterns/UnarySignPattern.java
package fi.quanfoxes.Parser.patterns;

import fi.quanfoxes.Lexer.*;
import fi.quanfoxes.Parser.Node;
import fi.quanfoxes.Parser.Pattern;
import fi.quanfoxes.Parser.ProcessedToken;
import fi.quanfoxes.Parser.nodes.ContextNode;
import fi.quanfoxes.Parser.nodes.NegateNode;
import fi.quanfoxes.Parser.nodes.NumberNode;

import java.util.List;

public class UnarySignPattern extends Pattern {
    private static final int PRIORITY = 14;

    private static final int OPERATOR = 0;
    private static final int SIGN = 1;
    private static final int OBJECT = 2;

    public UnarySignPattern() {
        super(TokenType.OPERATOR, TokenType.OPERATOR,
                TokenType.IDENTIFIER | TokenType.NUMBER | TokenType.PROCESSED);
    }

    @Override
    public int priority(List<Token> tokens) {
        return PRIORITY;
    }

    @Override
    public boolean passes(List<Token> tokens) {
        OperatorToken operator = (OperatorToken)tokens.get(OPERATOR);
        OperatorToken sign = (OperatorToken)tokens.get(SIGN);

        return operator.getOperator() != OperatorType.INCREMENT && operator.getOperator() != OperatorType.DECREMENT &&
                (sign.getOperator() == OperatorType.ADD || sign.getOperator() == OperatorType.SUBTRACT);
    }

    private Node getNode(Node parent, Token token) throws Exception {
        switch (token.getType()) {
            case TokenType.IDENTIFIER:
                ContextNode context = (ContextNode)parent;
                IdentifierToken identifier = (IdentifierToken)token;
                return context.getVariable(identifier.getIdentifier());
            case TokenType.NUMBER:
                NumberToken number = (NumberToken)token;
                return new NumberNode(number.getNumberType(), number.getNumber());
            case TokenType.PROCESSED:
                ProcessedToken process = (ProcessedToken)token;
                return process.getNode();
        }

        throw new Exception("INTERNAL_ERROR: Unhandled token");
    }

    @Override
    public Node build(Node parent, List<Token> tokens) throws Exception {
        OperatorToken sign = (OperatorToken)tokens.get(SIGN);
        Node object = getNode(parent, tokens.get(OBJECT));

        if (sign.getOperator() == OperatorType.ADD) {
            return object;
        }

        return new NegateNode(object);
    }

    @Override
    public int getStart(List<Token> tokens) {
        return super.getStart(tokens);
    }
}

File: Zigzag/Parser/Patterns/UnarySignPattern.cs

using System.Collections.Generic;

public class UnarySignPattern : Pattern
{
	public const int PRIORITY = 18;

	public const int OPERATOR = 0;
	public const int SIGN = 1;
	public const int OBJECT = 2;

	/// Example:
	/// a = -x
	public UnarySignPattern() : base
	(
		TokenType.KEYWORD | TokenType.OPERATOR | TokenType.OPTIONAL,
		TokenType.OPERATOR,
		TokenType.OBJECT
	) {}

	public override bool Passes(Context context, PatternState state, List<Token> tokens)
	{
		var sign = tokens[SIGN].To<OperatorToken>().Operator;
		return (sign == Operators.ADD || sign == Operators.SUBTRACT) && (!tokens[OPERATOR].Is(TokenType.NONE) || state.Start == 0);
	}

	public override Node? Build(Context context, List<Token> tokens)
	{
		var target = Singleton.Parse(context, tokens[OBJECT]);
		var sign = tokens[SIGN].To<OperatorToken>().Operator;

		if (target is NumberNode number)
		{
			if (sign == Operators.SUBTRACT)
			{
				number.Negate();
			}

			return number;
		}

		if (sign == Operators.SUBTRACT)
		{
			return new NegateNode(target);
		}

		return target;
	}

	public override int GetPriority(List<Token> tokens)
	{
		return PRIORITY;
	}

	public override int GetStart()
	{
		return SIGN;
	}
}